keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+
+#### Provide Clarity On Failure Scenarios
+
+This proposal introduces multiple failure scenarios that are going to be handled at the different places, and reported in different manner. Some will be caught during validation, other will be reported as an error event and will require troubleshooting by the user.  
+
+We'll have to provide clear documentation on the different failure scenarios, what the user can expect to be caught during valiation (eg storage class not supporting expansion) and what will be caught only during the reconciliation logic (eg user is trying to decrease size below or equal to PVC capacity; the CSI driver doesn't support online expansion). 
+
+## Implementation Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API changes
+
+The only API changes we'll make, is to add the `volumeClaimTemplates` array field to the StatefulSet status object as described in the [Feedback](#feedback) section.
+
+### API server validation relaxation
+Modifications would be made to `ValidateStatefulSetUpdate` in order to allow changes in the
+storage size of the volume (`.spec.resources.requests.storage`) in the `VolumeClaimTemplates`.
+
+### RBAC changes.
+Now that StatefulSet controller needs to patch the PVC, `patch` will be added to the `statefulset-controller` role in the buildControllerRoles.
+
+### StatefulSet controller changes
+
+#### PVC Capacity Notifications
+
+In order to be notified of capacity changes in PVC objects (arrow `(15)` in the diagram [here](#feedback)), which is necessary for the [feedback mechanism](#feedback), we'll perform the following changes
+
+* When [creating a new a PVC object during StatefulSet reconciliation](https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/controller/statefulset/stateful_pod_control.go#L342), we'll add an annotations to the PVC object, denoting the namespace and name of the StatefulSet object.
+  * We don't want to use the [`ownerReference`](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications)  in this case, because that field is already utilized by the [StatefulSet PVC retention mechanism](https://kubernetes.io/blog/2021/12/16/kubernetes-1-23-statefulset-pvc-auto-deletion/).
+* We'll start listening to PVC update (not add/delete) events in the [StatefulSet controller](https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/controller/statefulset/stateful_set.go#L80). When an update happens, we'll extract the StatefulSet object that we should do reconciliation for from the annotation we added in the previous step.
```

Comment:
so when kube-controller crashes and restarts, it will receive all updates it missed in the time it was down, as `PVC added` events? If so, then definetely, I'll have to address that

Reply:
Yes, KCM will get all PVCs as "PVC added" event when it starts. KCM will not get any "PVC updated" events for events that happened when it was down.